### PR TITLE
Include a hint on new default for form_with

### PR DESCRIPTION
**Pull request by NobodysNightmare**
### Summary

This Pull Request updates the upgrade instructions from Rails 5.2 to Rails 6.0 on using `form_with`.

In some private projects that I recently upgraded to Rails 6 I noticed that form submission didn't work properly after the upgrade. Cross-checking the [Form Helpers](https://guides.rubyonrails.org/form_helpers.html#how-do-forms-with-patch-put-or-delete-methods-work-questionmark) guide, I saw that it surely mentioned `remote: true` being the default for `form_with` and that one should use `local: true` to get "classic" form behaviour.

However, prior to the upgrade `form_with` seems to have submitted forms locally by default.

### Other Information

While writing this Pull Request I wanted to cross-check the [5.2 documentation](https://github.com/rails/rails/blob/5-2-stable/guides/source/form_helpers.md) on `form_with` and was surprised to not find it mentioned anywhere. Now I am wondering whether I was using an API that was not officially part of Rails 5.2 or whether it was just not documented in the guides back then.

Anyhow, `form_with` worked in Rails 5.2 and I might not be the only person that used it, so it's probably still worth including in the upgrade guide.

### CODE REVIEW

**Comment by jonathanhefner** [comment]
**To clarify(?) the issue:**

The default for `form_with` in Rails 5.1 was `local: false` i.e. remote = true (see [here](https://github.com/rails/rails/blob/v5.1.0/actionview/lib/action_view/helpers/form_helper.rb#L477) and [here](https://github.com/rails/rails/blob/v5.1.0/actionview/lib/action_view/helpers/form_helper.rb#L577-L578)).  But `new_framework_defaults_5_1.rb` included a setting for [remote = false](https://github.com/rails/rails/blob/v5.1.0/railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_1.rb.tt#L9-L10).  Notice that setting was uncommented though, unlike typical `new_framework_defaults_X_Y.rb` settings.  I think the intention was for users to comment this setting out when they "flip defaults one by one to migrate", and thus eventually arrive at remote = true.  The circumstances are a bit odd, though, since `form_with` was introduced in Rails 5.1, and so there would have been no existing `form_with` usages for that setting to affect.

Also note that Rails 5.2 [added that setting](https://github.com/rails/rails/blob/v5.2.0/railties/lib/rails/application/configuration.rb#L87-L89) to `load_defaults` since [it was omitted](https://github.com/rails/rails/blob/v5.1.0/railties/lib/rails/application/configuration.rb#L73-L79) in Rails 5.1.

**To sum up:**

I don't believe this new section is correct.  Can you verify what your `form_with_generates_remote_forms` setting was before the upgrade?

> I wanted to cross-check the [5.2 documentation](https://github.com/rails/rails/blob/5-2-stable/guides/source/form_helpers.md) on `form_with` and was surprised to not find it mentioned anywhere.

I think that was just an oversight.  It was at least [mentioned](https://guides.rubyonrails.org/v5.1/getting_started.html#the-first-form) in Rails 5.1 Getting Started Guide.

**Comment by NobodysNightmare** [comment]
> Can you verify what your form_with_generates_remote_forms setting was before the upgrade?

On the rails console I tried `Rails.application.config.form_with_generates_remote_forms` for both the Rails 5.2 and the Rails 6.0 version of my application. In both versions it raises a `NoMethodError`.

Just to be sure:
* I don't have a `new_framework_defaults_*.rb` file in my `config/initializers`
* I couldn't find an explicit mention of `form_with_generates_remote_forms` in my code base

However, I noticed that I didn't have the `config.load_defaults` line at all in my `config/application.rb` until my migration to Rails 6, where I added it manually. This might have been an error on my side regarding the upgrade procedure (though I am not sure when `load_defaults` was introduced).

In case you want to see my app config in general, it is [here](https://github.com/NobodysNightmare/sprit/commit/21275c8a2796b1a4101c475ce51866336ebeb4f6):

* The app started with a `rails new` (as of 5.0) template
* Then upgrade to 5.2
* Then upgrade to 6.0
* Honestly I think I left out the `app:update` rake task on upgrades, so this might have made my config deviate from normal expectations

**Comment by jonathanhefner** [comment]
> On the rails console I tried `Rails.application.config.form_with_generates_remote_forms` for both the Rails 5.2 and the Rails 6.0 version of my application. In both versions it raises a `NoMethodError`.

Try `Rails.application.config.action_view.form_with_generates_remote_forms` instead.  You could also verify that `ActionView::Helpers::FormHelper.form_with_generates_remote_forms` matches that value, just in case.

> However, I noticed that I didn't have the `config.load_defaults` line at all in my `config/application.rb` until my migration to Rails 6, where I added it manually.

If you go back and try the 5.2 version of your app with `load_defaults 5.2`, does it behave the same (with respect to `form_with`) as the 6.0 version with `load_defaults 6.0`?

> * The app started with a `rails new` (as of 5.0) template
> 
> * Then upgrade to 5.2

(In case this wasn't a typo) You should always upgrade to the next minor release, i.e. 5.0 -> 5.1 -> 5.2 -> 6.0.  It helps to ensure a smooth upgrade experience.

> * Honestly I think I left out the `app:update` rake task on upgrades, so this might have made my config deviate from normal expectations

`app:update` is what generates `new_framework_defaults.rb`.  Using `new_framework_defaults.rb` [as recommended](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#configure-framework-defaults) also helps to ensure a smooth upgrade experience.

**Comment by NobodysNightmare** [comment]
> Try `Rails.application.config.action_view.form_with_generates_remote_forms` instead

This works. Using my "vanilla" 5.2 app it returns `nil`. After I added `config.load_defaults '5.2'` to the `application.rb` it was still `nil`. **However**, `ActionView::Helpers::FormHelper.form_with_generates_remote_forms` changes from `nil` to `true` when I add the `load_defaults`, so this was apparently the culprit here.

Regarding the smooth upgrade experience:

Yeah, this wasn't a typo. I just wanted to be open and clear about the points where I worked "against" the documentation, because I didn't consider this being caused by my upgrade behaviour before your first response.

My assumption so far was, that **all** necessary changes were documented in the Upgrading from Rails X to Rails Y sections. I wanted to "skip" the `new_framework_defaults` file and just do the entire upgrade in one step so I thought I'd just need to read through all of the changes in the doc (which mostly didn't affect my application). I think this was a flawed assumption.

As far as I can now tell, the docs just talk about changes that are highly likely to be breaking, but more changes (with smoother options) might be breaking, but are feathered (and documented) in the `new_framework_defaults` files.

Long story short: I guess my PR can be closed, thanks for helping me understand the underlying issue. Maybe I was able to provide some input for why I did the upgrade in the wrong way ^^
Let me know in case I can do anything regarding my PR to still be useful for the project.

Thanks a lot.

**Comment by jonathanhefner** [comment]
> Long story short: I guess my PR can be closed, thanks for helping me understand the underlying issue. Maybe I was able to provide some input for why I did the upgrade in the wrong way ^^
> Let me know in case I can do anything regarding my PR to still be useful for the project.

Glad I could help!  PRs which improve the documentation are always welcome, if there are changes that you feel would have prevented the confusion. :smile:
